source/Nevermore/Advanced/Queryable/QueryProvider.cs

```diff
@@ -86,7 +87,27 @@ public async Task<TResult> ExecuteAsync<TResult>(Expression expression, Cancella
             {
                 var asyncStream = (IAsyncEnumerable<object>)GenericStreamAsyncMethod.MakeGenericMethod(expression.Type)
                     .Invoke(queryExecutor, new object[] { command, cancellationToken });
-                return (TResult)Convert.ChangeType(await FirstOrDefaultAsync(asyncStream, cancellationToken), typeof(TResult));
+                var firstOrDefaultAsync = await FirstOrDefaultAsync(asyncStream, cancellationToken);
+
+                if (firstOrDefaultAsync is null)
+                {
+                    if (typeof(TResult).IsValueType)
+                    {
+                        throw new InvalidCastException("Null object cannot be converted to a value type");
+                    }
+
+                    // TODO: This NEEDS to go away when we turn nullable on in Nevermore
+                    // This method needs to be able to return null for instances like `FirstOrDefaultAsync` 
+                    object GetNull() => null;
+                    return (TResult) GetNull();
+                }
+                
+                if (!firstOrDefaultAsync.GetType().IsAssignableFrom(typeof(TResult)))
+                {
+                    throw new InvalidCastException($"Type {firstOrDefaultAsync.GetType()} is not castable to type {typeof(TResult)}");
+                }
```

Comment:
This check is replicating what you get for free by just doing the type cast below.